packages/experimental-utils/src/ast-utils/predicates.ts

```diff
@@ -7,6 +7,28 @@ const isNodeOfType =
   ): node is TSESTree.Node & { type: NodeType } =>
     node?.type === nodeType;
 
+type ObjectEntry<BaseType> = [keyof BaseType, BaseType[keyof BaseType]];
+type ObjectEntries<BaseType> = Array<ObjectEntry<BaseType>>;
+const isNodeOfTypeWithConditions = <
+  NodeType extends AST_NODE_TYPES,
+  Conditions extends Partial<TSESTree.Node & { type: NodeType }>,
+>(
+  nodeType: NodeType,
+  conditions: Conditions,
+): ((
+  node: TSESTree.Node | null | undefined,
+) => node is TSESTree.Node & { type: NodeType } & Conditions) => {
```

Comment:
Due to the extraction of `entries` & `@typescript-eslint/explicit-function-return-type`, we have to duplicate this return type

Reply:
I'm fine with the duplication on the util.
I'm also happy to just disable the linter on this check too.
We can merge this as is.